libraries/AP_Arming/AP_Arming.cpp

```diff
@@ -435,15 +434,49 @@ bool AP_Arming::gps_checks(bool report)
             return false;
         }
 
-        // check AHRS and GPS are within 10m of each other
         const Location gps_loc = gps.location();
         Location ahrs_loc;
         if (AP::ahrs().get_position(ahrs_loc)) {
+            // check AHRS and GPS are within 10m of each other horizontally
             const float distance = gps_loc.get_distance(ahrs_loc);
-            if (distance > AP_ARMING_AHRS_GPS_ERROR_MAX) {
+            const uint8_t AHRS_GPS_ERROR_MAX = 10;      // accept up to 10m difference between AHRS and GPS horizontally
+            if (distance > AHRS_GPS_ERROR_MAX) {
                 check_failed(ARMING_CHECK_GPS, report, "GPS and AHRS differ by %4.1fm", (double)distance);
                 return false;
             }
+            // check AHRS and GPS are close vertically
+            int32_t gps_alt_cm;
+            if (!gps_loc.get_alt_cm(Location::AltFrame::ABSOLUTE, gps_alt_cm)) {
+                check_failed(ARMING_CHECK_GPS, report, "Failed to get %s altitude", "GPS");
+                return false;
+            }
+            int32_t ahrs_alt_cm;
+            if (!ahrs_loc.get_alt_cm(Location::AltFrame::ABSOLUTE, ahrs_alt_cm)) {
+                check_failed(ARMING_CHECK_GPS, report, "Failed to get %s altitude", "AHRS");
+                return false;
+            }
+            const uint32_t distance_z_cm = labs(gps_alt_cm - ahrs_alt_cm);
+            const uint16_t AHRS_GPS_ERROR_Z_MAX_CM = 10000;      // accept up to 100m difference between AHRS and GPS vertically
+            if (distance_z_cm > AHRS_GPS_ERROR_Z_MAX_CM) { //
+                check_failed(ARMING_CHECK_GPS, report, "Bad GPS/AHRS alt delta %.1fm", distance_z_cm/100.0f);
+                return false;
+            }
+            if ((checks_to_perform & ARMING_CHECK_ALL) || (checks_to_perform & ARMING_CHECK_BARO)) {
+                int32_t gps_alt_rel_cm;
+                if (!gps_loc.get_alt_cm(Location::AltFrame::ABOVE_HOME, gps_alt_rel_cm)) {
+                    check_failed(ARMING_CHECK_GPS, report, "Failed to retrieve GPS home-relative altitude");
+                    return false;
+                }
+#if !APM_BUILD_TYPE(APM_BUILD_APMrover2) // Rover doesn't get altitudes correct
+                // also check that the Baro and GPS align:
+                const uint16_t BARO_GPS_ERROR_Z_MAX_CM = 5000;      // accept up to 50m difference between BARO and GPS vertically
```

Comment:
The compilier should be fixing this for us, but why declare this as `uint16_t`?  You only use it where it will be implicitly promoted to float. In fact if you make this a float in meters, then you just have to multiple `gps_alt_rel_cm` by `1e-2f` and you have exactly the same number of operations  if the check passes, one less divide if it fails, and you have sane units, and the define gets way easier to read.

Reply:
Fixed.